### PR TITLE
PT-4708: Download personal data for specific customer

### PR DESCRIPTION
**Pull request by einlied753**

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?><Project>
   <!-- These properties will be shared for all projects -->
   <PropertyGroup>
-    <VersionPrefix>1.1.0</VersionPrefix>
+    <VersionPrefix>1.2.0</VersionPrefix>
```
Comment (Vectorfield4) [inline Directory.Build.props:4]: Revert these changes.

File: src/VirtoCommerce.GDPR.Core/Models/DownloadData/Account.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+namespace VirtoCommerce.GDPR.Core.Models.DownloadData
+{
+    public class Account
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Core/Models/DownloadData/Account.cs:3]: Add /// description

File: src/VirtoCommerce.GDPR.Core/Models/DownloadData/Address.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+namespace VirtoCommerce.GDPR.Core.Models.DownloadData
+{
+    public class Address
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Core/Models/DownloadData/Address.cs:3]: Add /// description

File: src/VirtoCommerce.GDPR.Core/Models/DownloadData/Customer.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+using System.Collections.Generic;
+
+namespace VirtoCommerce.GDPR.Core.Models.DownloadData
+{
+    public class Customer
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Core/Models/DownloadData/Customer.cs:5]: Add /// description

File: src/VirtoCommerce.GDPR.Core/Models/DownloadData/Customer.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public string FirstName { get; set; }
+        public string LastName { get; set; }
+        public string FullName { get; set; }
+        public System.DateTime? Birthday { get; set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Core/Models/DownloadData/Customer.cs:10]:
```suggestion
        public DateTime? Birthday { get; set; }
```

File: src/VirtoCommerce.GDPR.Core/Services/IDownloadContactDataService.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+using VirtoCommerce.GDPR.Core.Models.DownloadData;
+
+namespace VirtoCommerce.GDPR.Core.Services
+{
+    public interface IDownloadContactDataService
+    {
+        Task<Customer> GetContactData(string id);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Core/Services/IDownloadContactDataService.cs:8]:
```suggestion
        Task<Customer> GetContactDataAsync(string id);
```

File: src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+using VirtoCommerce.Platform.Core.GenericCrud;
+
+namespace VirtoCommerce.GDPR.Data.Services
+{
+    public class DownloadContactDataService : IDownloadContactDataService
+    {
+        private readonly IMemberService _memberService;
+        private readonly ISearchService<CustomerOrderSearchCriteria, CustomerOrderSearchResult, CustomerOrder> _customerOrderSearchService;
+
+        public DownloadContactDataService(IMemberService memberService, ICustomerOrderSearchService customerOrderSearchService)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs:20]: Inject `ISearchService<CustomerOrderSearchCriteria, CustomerOrderSearchResult, CustomerOrder>` instead of `ICustomerOrderSearchService`, remove casting.

File: src/VirtoCommerce.GDPR.Web/module.manifest [hunk 8]
```diff
@@ -1,32 +1,25 @@
...
-            <permission id="virtoCommerce.gdpr:access" name="Access" />
-            <permission id="virtoCommerce.gdpr:delete" name="Delete" />
-        </group>
-    </permissions>
-    <releaseNotes>First version.</releaseNotes>
-    <copyright>Copyright © 2021 VirtoCommerce. All rights reserved</copyright>
-    <tags>extension module</tags>
-</module>
+  <id>VirtoCommerce.GDPR</id>
+  <version>1.2.0</version>
```
Comment (Vectorfield4) [inline src/VirtoCommerce.GDPR.Web/module.manifest:4]: Revert these changes.

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        public async Task<Customer> GetContactData(string id)
+        {
+            Contact contact = (Contact)await _memberService.GetByIdAsync(id);
+
+            var orderSearchCriteria = AbstractTypeFactory<CustomerOrderSearchCriteria>.TryCreateInstance();
+            orderSearchCriteria.CustomerId = contact.Id;
+            orderSearchCriteria.Take = 25;
+            var customerOrders = await _customerOrderSearchService.SearchAsync(orderSearchCriteria);
+
+            Customer result = new Customer()
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs:40]: var result

File: src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        /// </summary>
+        /// <param name="id">contact Id</param>
+        /// <returns></returns>
+        public async Task<Customer> GetContactData(string id)
+        {
+            Contact contact = (Contact)await _memberService.GetByIdAsync(id);
+
+            var orderSearchCriteria = AbstractTypeFactory<CustomerOrderSearchCriteria>.TryCreateInstance();
+            orderSearchCriteria.CustomerId = contact.Id;
+            orderSearchCriteria.Take = 25;
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs:37]: Add a comment why we take only 25 orders

File: src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        /// <summary>
+        /// Returns personal contact information
+        /// </summary>
+        /// <param name="id">contact Id</param>
+        /// <returns></returns>
+        public async Task<Customer> GetContactData(string id)
+        {
+            Contact contact = (Contact)await _memberService.GetByIdAsync(id);
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs:33]: var contact
Don't use strict casting, use `as Contact` or even base Member type, if we don't use specific Contact type fields

File: src/VirtoCommerce.GDPR.Core/VirtoCommerce.GDPR.Core.csproj [hunk 12]
```diff
@@ -10,14 +10,14 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="VirtoCommerce.CustomerModule.Core" Version="3.34.0" />
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Core/VirtoCommerce.GDPR.Core.csproj:13]: Add these dependencies to the module.manifest

File: src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs (hunk 10 above)
Comment (Vectorfield4) [reply src/VirtoCommerce.GDPR.Data/Services/DownloadContactDataService.cs:37]: https://github.com/VirtoCommerce/vc-module-experience-api/blob/dev/src/XPurchase/VirtoCommerce.XPurchase/Services/CartProductService.cs#L30

File: src/VirtoCommerce.GDPR.Web/Controllers/Api/GdprController.cs [hunk 13]
```diff
@@ -52,12 +55,27 @@ public async Task<ActionResult<Contact[]>> GetContactList([FromBody] MembersSear
...
+        /// <summary>
+        /// Download contact info by id
+        /// </summary>
+        /// <param name="id">contact id</param>
+        /// <returns></returns>
+        [HttpGet]
+        [Route("contacts/download")]
+        [Authorize(ModuleConstants.Security.Permissions.Download)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        public async Task<ActionResult> DownloadContactInfo([FromQuery] string id)
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Web/Controllers/Api/GdprController.cs:73]:
```suggestion
        public async Task<ActionResult<Customer>> DownloadContactInfo([FromQuery] string id)
```

File: src/VirtoCommerce.GDPR.Web/Controllers/Api/GdprController.cs [hunk 14]
```diff
@@ -52,12 +56,27 @@ public async Task<ActionResult<Contact[]>> GetContactList([FromBody] MembersSear
...
 
+        /// <summary>
+        /// Download contact info by id
+        /// </summary>
+        /// <param name="id">contact id</param>
+        /// <returns></returns>
+        [HttpGet]
+        [Route("contacts/download")]
+        [Authorize(ModuleConstants.Security.Permissions.Download)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
```
Comment (pushnitsa) [inline src/VirtoCommerce.GDPR.Web/Controllers/Api/GdprController.cs:73]:
```suggestion
        [ProducesResponseType(typeof(Customer), StatusCodes.Status204NoContent)]
```

**Review by pushnitsa — APPROVED** [review]

**Review by Vectorfield4 — APPROVED** [review]
